File: src/faces/stopwatch.cpp

#include "./faces/stopwatch.h"

#include "config.h"
#include "gfx_util.h"
#include "osw_app.h"
#include "osw_hal.h"
#include "osw_ui_util.h"

// continue after sleep does not work yet
// because millis restarts from 0
// TODO: use the I2C RTC instead
RTC_DATA_ATTR long start = 0;
RTC_DATA_ATTR long diff = 0;
RTC_DATA_ATTR bool running = false;
RTC_DATA_ATTR bool reset = true;
RTC_DATA_ATTR long sumPaused = 0;

// OswAppHelloWorld::OswAppHelloWorld(void) : OswApp() {}
void OswStopwatchFace::setup(OswLauncher*launcher ) {
  _launcher = launcher;
}

void OswStopwatchFace::action(){
     if (reset) {  // Start
      start = _launcher->getHal()->getLocalTime();
    } else {  // Continue
      sumPaused += diff;
      start = _launcher->getHal()->getLocalTime();
    }
    running = true;
    reset = false;
    _launcher->getHal()->clearBtn3();
}
void OswStopwatchFace::action2(){
 if (running) {  // Stop
      running = false;
    } else {  // Reset
      diff = 0;
      sumPaused = 0;
      reset = true;
    }
    _launcher->getHal()->clearBtn2();
}

void OswStopwatchFace::loop() {
  OswHal* hal = _launcher->getHal();
  hal->getCanvas()->fillScreen(0);
  hal->getCanvas()->setTextColor(rgb565(255, 255, 255));

  hal->getCanvas()->setTextSize(2);
  if (reset) {
    hal->getCanvas()->setCursor(220 - defaultFontXOffset(5, 2), 42);
    hal->getCanvas()->print("Start");
  } else if (!running) {
    hal->getCanvas()->setCursor(220 - defaultFontXOffset(8, 2), 42);
    hal->getCanvas()->print("Continue");
  }

  if (running) {
    hal->getCanvas()->setCursor(220 - defaultFontXOffset(4, 2), 182);
    hal->getCanvas()->print("Stop");
  } else if (!reset) {
    hal->getCanvas()->setCursor(220 - defaultFontXOffset(5, 2), 182);
    hal->getCanvas()->print("Reset");
  }

  if (running) {
    diff = hal->getLocalTime() - start;
  }

  long total = diff + sumPaused;
  long deltaSeconds = (total) % 60;
  long deltaMinutes = (total / 60) % 60;
  long deltaHours = (total / 60 / 60) % 24;
  long deltaDays = total / 60 / 60 / 24;

    if (deltaDays) {
      hal->gfx()->setTextSize(4);
      hal->gfx()->setTextBottomAligned();
      hal->gfx()->setTextCenterAligned();
      hal->gfx()->setTextCursor(120, 120);
      hal->gfx()->print((String(deltaDays) + "d").c_str());
    }

    hal->gfx()->setTextSize(4);
    if (deltaDays) {
      hal->gfx()->setTextTopAligned();
    } else {
      hal->gfx()->setTextMiddleAligned();
    }
    hal->gfx()->setTextLeftAligned();
    // manually center the counter:
    hal->gfx()->setTextCursor(120 - hal->gfx()->getTextOfsetColumns(4), 120);
    hal->gfx()->printDecimal(deltaHours, 2);
    hal->gfx()->print(":");
    hal->gfx()->printDecimal(deltaMinutes, 2);
    hal->gfx()->print(":");
    hal->gfx()->printDecimal(deltaSeconds, 2);

  _launcher->requestFlush();
}

void OswStopwatchFace::stop() {
  // hal->disableDisplayBuffer();
}

File: data/lua/apps/stopwatch.lua
graphicsUtils = require("graphicsUtils")

function defaultFontXOffset(numChars, scale)  --works with default font only
    return numChars * 6 * scale;
end

function defaultFontYOffset(numRows, scale)
    return numRows * 8 * scale;
end


start = 0;
diff = 0;
running = false;
reset = true;
sumPaused = 0;

function setup()
    print("Started stop watch")
end

function loop()
    if (hal:btnHasGoneDown(osw.BUTTON_3)) then
        if (reset == true) then
          start = hal:getLocalTime();
        else
          sumPaused = sumPaused + diff;
          start = hal:getLocalTime();
        end
        running = true;
        reset = false;
    end
    if (hal:btnHasGoneDown(osw.BUTTON_2)) then
        if (running) then
            running = false
        else
            diff = 0
            sumPaused = 0
            reset = true
        end
    end

    hal:getCanvas():fillScreen(0);
    hal:getCanvas():setTextColor(graphicsUtils.rgb565(255, 255, 255));

    hal:getCanvas():setTextSize(2);
    if (reset == true) then
        hal:getCanvas():setCursor(220 - defaultFontXOffset(5, 2), 42);
        hal:getCanvas():print("Start");
    elseif (running == false) then
        hal:getCanvas():setCursor(220 - defaultFontXOffset(8, 2), 42);
        hal:getCanvas():print("Continue");
    end

    if (running == true) then
        hal:getCanvas():setCursor(220 - defaultFontXOffset(4, 2), 182);
        hal:getCanvas():print("Stop");
    elseif (reset == false) then
        hal:getCanvas():setCursor(220 - defaultFontXOffset(5, 2), 182);
        hal:getCanvas():print("Reset");
    end

    if (running == true) then
        diff = hal:getLocalTime() - start;
    end

    total = math.floor(diff + sumPaused);
    deltaSeconds = (total) % 60;
    deltaMinutes = (total // 60) % 60;
    deltaHours = (total / 60 // 60) % 24;
    deltaDays = total / 60 / 60 // 24;

    hal:getCanvas():setTextSize(4);

    if (deltaDays > 0) then
        hal:getCanvas():setCursor(120 - defaultFontXOffset(deltaDays < 10 and 1 or 2, 4),
                                    120 - defaultFontYOffset(1, 4) * 1.5);
        hal:getCanvas():print(string.format("%dd", deltaDays));
    end

    hal:getCanvas():setCursor(defaultFontXOffset(1, 4), 120 - defaultFontYOffset(1, 4) / 2);
    hal:getCanvas():print(string.format("%02d:%02d:%02d", deltaHours, deltaMinutes, deltaSeconds))

    hal:flushCanvas()
end

function stop()
end
